File: app/web/ui/tryck.js
/* ══════════ UTSKRIFTSPAKETET ══════════
   «Det här ska skrivas ut» som en enda gest. Appen vet vad lektionen består av
   och packar högen själv: tavlan överst, elevernas papper under, facit sist.
   Anpassade kopior (post 11) är rader i paketet — en egenskap hos KOPIAN, aldrig
   en etikett på pappret. Sammanräkningen står i knappen, före utskriften. */
(() => {
  const $ = s => document.querySelector(s);
  const ruta = $('#tryckruta');
  if (!ruta) return;

  const SIDOR = { Tavla: 1, Prov: 3, Arbetsblad: 2, Gruppuppgift: 1, Anteckningar: 1 };
  /* Sidantalet ska vara bladens, inte en tabell som gissar: formen vet exakt hur
     många ark dokumentet består av. En tavla är ett ark — den stod som två. */
  const sidorFor = v => {
    /* Tavlan är ETT papper i förhandsvisningen men ett bräde per sida i
       utskriften — räkningen ska vara högens, inte skärmens. */
    if (v.typ === 'Tavla' && window.TavlaBild) return window.TavlaBild.antal(v);
    const f = window.Blad && window.Blad.form ? window.Blad.form(v) : null;
    return (f && f.length) || SIDOR[v.typ] || 2;
  };
  let rader = [], titel = '';

  function packa(namnPaPaketet, dokument) {
    titel = namnPaPaketet;
    /* Anteckningarna ligger överst i högen, före tavlan: det är pappret läraren
       håller i medan hon säger det som står på tavlan. */
    const ordning = { Anteckningar: -1, Tavla: 0, Prov: 1, Arbetsblad: 1, Facit: 2 };
    /* Elevmaterialet trycks i klassuppsättning, lärarens papper i ETT exemplar.
       Anteckningarna är lärarens — tjugotvå kopior av hennes stödpapper är
       tjugoett papper i papperskorgen. */
    const LARARENS = { Tavla: 1, Anteckningar: 1 };
    rader = dokument.map(v => ({
      v,
      namn: window.Dokument.namn(v),
      typ: v.losningsblad ? 'Facit' : v.typ,
      under: v.losningsblad
        ? 'Din kopia · enkelsidig'
        : v.typ === 'Tavla' ? 'Din genomgång · enkelsidig'
        : v.typ === 'Anteckningar' ? 'Ditt stödpapper · enkelsidig'
        : 'Standard · dubbelsidig',
      antal: v.losningsblad ? 1 : LARARENS[v.typ] || 22,
      sidor: sidorFor(v),
      med: true
    })).sort((a, b) => (ordning[a.typ] ?? 1) - (ordning[b.typ] ?? 1));
    /* Formelbladet är en bilaga i provet — samma dokument, sista sidan — men
       den kan skrivas ut med eller utan i paketet. Ingen fjärde dokumenttyp. */
    const prov = rader.find(r => r.typ === 'Prov'
                                 && r.v.inst && r.v.inst.formelblad);
    if (prov) rader.splice(rader.indexOf(prov) + 1, 0, {
      v: prov.v, namn: 'Formelblad · ' + (prov.v.kurs || 'kursen'), typ: 'Bilaga',
      under: `Bilaga, sida ${sidorFor(prov.v) + 1} — härledd ur uppgifterna, skrivs ut med provet`,
      antal: prov.antal, sidor: 1, med: true
    });
    /* Lösningsförslaget till BOKENS uppgifter är lärarens eget ark, inte ett
       facit till uppgifter appen skrivit — det hör därför sist i högen, där
       facit ligger, och räknas som en kopia. Urvalet är dokumentets egen lista
       (v.bokuppg), samma som tavlans vänsterspalt skriver upp. */
    rader.filter(r => !r.v.losningsblad && r.v.bokuppg && r.v.bokuppg.losning && r.v.bokuppg.losning.antal)
      .forEach(r => {
        const l = r.v.bokuppg.losning;
        /* Antalet är de uppgifter arket FAKTISKT bär en lösning till, inte hela
           urvalet. Allt på olästa sidor — och det modellen hoppade över —
           kommer tillbaka utan innehåll, och BokLosning ritar inte tomma
           poster. Raden lovade förr hela urvalet, så kvittot i handen sa en
           annan sak än pappret i skrivaren. Utan server (Claude Design) är
           posterna prototypmallar och alla ritas: då gäller urvalet. */
        const skrivna = (window.API && window.API.pa)
          ? (l.poster || []).filter(p => p.text && p.svar).length
          : l.antal;
        const raknat = skrivna === l.antal
          ? `${l.antal} ${l.antal === 1 ? 'uppgift' : 'uppgifter'}`
          : `${skrivna} av ${l.antal} uppgifter`;
        rader.push({
          v: r.v, bok: true,
          namn: `Lösningsförslag · boken s. ${r.v.bokuppg.sidor}`, typ: 'Facit',
          under: `${raknat} (${l.niva.toLowerCase()}) · uppg ${l.remsa} · din kopia · enkelsidig`,
          /* Sidorna är arken BokLosning faktiskt sätter — svarsfacit, bedömd
             elevlösning per nivå 3-uppgift, nivå 3-facit — inte en gissning ur
             uppgiftsantalet. Paginerar ett ark sig blir det ändå en sida till;
             raden är alltså en undre gräns, och kvittot räknar det riktiga. */
          antal: 1, med: true,
          sidor: Math.max(1, (window.BladBild && window.BladBild.antal(r.v))
                              || Math.ceil(Math.max(skrivna, 1) / 4))
        });
      });
    rita();
    ruta.hidden = false;
  }

  function rita() {
    $('#trycktitel').textContent = 'Att skriva ut · ' + titel;
    const lista = $('#trycklista');
    lista.innerHTML = '';
    rader.forEach((r, i) => {
      const rad = document.createElement('div');
      rad.className = 'tryckrad';
      rad.innerHTML = `<span class="refminis" data-typ="${r.typ}">${r.typ[0]}</span><span><span class="trycknamn"></span><span class="tryckunder"></span></span><span class="tryckantal">${r.antal} ex · ${r.sidor} sid</span><button class="lank" type="button">${r.med ? 'Utelämna' : 'Ta med'}</button>`;
      rad.querySelector('.trycknamn').textContent = r.namn;
      rad.querySelector('.tryckunder').textContent = r.under;
      rad.style.opacity = r.med ? '1' : '.45';
      rad.querySelector('button').addEventListener('click', () => { rader[i].med = !rader[i].med; rita(); });
      lista.appendChild(rad);
    });
    const med = rader.filter(r => r.med);
    const sidor = med.reduce((a, r) => a + r.antal * r.sidor, 0);
    $('#tryckskicka').textContent = `Skriv ut ${med.length} ${med.length === 1 ? 'dokument' : 'dokument'} · ${sidor} sidor`;
    $('#tryckskicka').disabled = !med.length;
  }

  /* Anpassad kopia: förlängd tid, färre uppgifter, luftigare sättning. Väljs
     per kopia, syns bara i sättningen och i dokumentkoden i foten. */
  $('#tryckanpassad').addEventListener('click', () => {
    const bas = rader.find(r => r.typ === 'Prov' || r.typ === 'Arbetsblad');
    if (!bas) return;
    if (rader.some(r => r.anpassad)) { window.toast && window.toast('Det finns redan en anpassad kopia i paketet'); return; }
    const i = rader.indexOf(bas);
    rader.splice(i + 1, 0, {
      v: bas.v, namn: bas.namn, typ: bas.typ, anpassad: true,
      under: '4 uppgifter · 150 min · luftigare sättning, 13 pt · koden i foten skiljer den',
      antal: 2, sidor: bas.sidor, med: true
    });
    bas.antal = Math.max(1, bas.antal - 2);
    rita();
    window.toast && window.toast('Anpassad kopia tillagd — ingen etikett på pappret', 'Ångra', () => {
      rader = rader.filter(r => !r.anpassad);
      bas.antal += 2;
      rita();
    });
  });

  /* ══════════ PAKETET PÅ RIKTIGT ══════════
     Knappen räknade ihop högen och sa «Utskrivet» efter niohundra
     millisekunder. Servern bygger den nu: provets och arbetsbladets PDF:er
     (byggda vid godkännandet), tavlan som bild, facit bredvid provet — i den
     här ordningen, med kopiorna I FILEN. En lärare som ska ha 22 elevark, en
     tavla och ett facit kan inte säga det i en skrivardialog som har ETT
     kopieantal för hela jobbet.

     Utan server spelas prototypens kvittering upp precis som förut. */
  const serverPa = () => !!(window.API && window.API.pa);
  /* Tavlan finns bara som ritad DOM i webbläsaren — servern kan inte rendera om
     den, för motorn bor här och inte där. Den ritas därför av på klienten
     (tavla-bild.js) och skickas med som PNG; servern lägger bilden på ett A4
     överst i paketet. Går avritningen inte igenom hamnar tavlan i `saknas` och
     kvittot säger det, i stället för att paketet tyst blir en sida kortare. */

  function paketkropp() {
    const med = rader.filter(r => r.med);
    return Promise.all(med.map(r => {
      const d = { namn: r.namn, typ: r.typ, kopior: r.antal };
      /* Lösningsförslaget till BOKENS uppgifter ritas bara i webbläsaren
         (BokLosning) — det finns ingen byggd fil på servern. Raden delar
         dokument med sitt original, så utan den här avfarten skulle den få
         originalets EGEN pdf under bokens namn.
         Förr slutade avfarten här: utan id hamnade raden i `saknas`, och
         kvittot sa det. Sant, men inte till någon nytta — läraren såg arken på
         skärmen och fick dem aldrig på papper. Nu ritas de av precis som
         tavlan (blad-bild.js) och skickas som en lista PNG:er, ett ark per
         sida. `saknas`-vägen står kvar som fallback: går avritningen inte
         igenom skickas ingen bild, och kvittot säger det igen. */
      if (r.bok) {
        if (!window.BladBild) return d;
        return window.BladBild.boklos(r.v)
          .then(ark => (ark.length ? Object.assign(d, { png: ark.map(a => a.png) }) : d))
          .catch(() => d);
      }
      /* Anteckningarna ligger i samma tabell som proven på servern, så
         paketets exam-gren hämtar deras PDF utan en rad ny kod — men utan
         bedömning och utan anpassad kopia: det finns ingen bedömning att lägga
         bredvid, och en anpassad kopia av lärarens eget papper är ingenting. */
      /* `provBorta`: provraden är raderad (en ångrad radering i planeringen).
         Id:t duger då inte att hämta filer med — radnummer återanvänds — och
         raden hamnar i `saknas`, vilket är sant: det finns ingen PDF. */
      if (r.v && r.v.provBorta) return d;
      if (r.v && r.v.antId) d.exam_id = r.v.antId;
      if (r.v && r.v.provId) {
        d.exam_id = r.v.provId;
        /* Provets lösningsförslag är det avritade facitläget; arbetsbladets är
           det separata facit. Två skilda filer, båda bredvid dokumentets egen
           PDF. Raden bad förut om bedömningen för båda: arbetsbladets facit
           hamnade därför alltid i `saknas`, och provet fick lärarens
           rättningsdokument i stället för arket hon såg på skärmen. */
        if (r.typ === 'Facit') d[r.v.typ === 'Prov' ? 'losningar' : 'facit'] = true;
        if (r.anpassad) d.anpassad = {
          /* Samma anpassning som raden beskriver: förlängd tid, färre
             uppgifter, och en kod i foten som skiljer kopian. */
          tid_min: 150, antal: 4,
          kod: `${(r.v.klass || 'kopia')}-${(r.v.datum || '').slice(5)}`,
        };
      }
      if (r.typ !== 'Tavla' || !window.TavlaBild) return d;
      /* Ett bräde per sida, som en lista — samma väg som bokens ark. Hela
         remsan i en bild blev en rand mitt på ett A4; nu fyller varje bräde
         sin egen liggande sida (tryck.png_till_pdf väljer orienteringen). */
      return window.TavlaBild.sidor(r.v)
        .then(png => Object.assign(d, { png }))
        .catch(() => d);      /* utan bild går raden till `saknas` — och sägs */
    })).then(dokument => ({ titel, dokument }));
  }

  function bygg(knapp, tillagg, efterat) {
    const text = knapp.textContent;
    knapp.disabled = true;
    /* Avritningen av tavlan tar sina hundradelar och sker före anropet —
       knappen ska säga vad den gör, inte stå tyst. */
    knapp.textContent = rader.some(r => r.med && (r.typ === 'Tavla' || r.bok))
      ? 'Ritar av …' : 'Bygger paketet …';
    const ater = () => { knapp.textContent = text; knapp.disabled = false; };
    paketkropp().then(kropp => {
      knapp.textContent = 'Bygger paketet …';
      return window.API.strom('/api/tryck', Object.assign(kropp, tillagg), {
        log: m => { knapp.textContent = String(m || '').slice(0, 40); },
      });
    }).then(res => {
      if (!res) throw new Error('Servern slutade svara mitt i bygget.');
      efterat(res);
      ater();
    }).catch(e => {
      ater();
      window.toast && window.toast(e.message || 'Paketet gick inte att bygga.');
    });
  }

  $('#tryckskicka').addEventListener('click', () => {
    const b = $('#tryckskicka');
    if (!serverPa()) {
      const text = b.textContent;
      b.disabled = true;
      b.textContent = 'Skickar till skrivaren …';
      setTimeout(() => {
        b.textContent = 'Utskrivet';
        window.toast && window.toast('Papperen ligger i rätt ordning i utskriftskön');
        setTimeout(() => { b.textContent = text; b.disabled = false; }, 1600);
      }, 900);
      return;
    }
    /* «Skriv ut» förblir EN hopfogad fil — det är hela poängen med högen:
       kopiorna ligger i den och bunten är rätt när den kommer ur maskinen. */
    bygg(b, {}, res => {
      /* Paketet öppnas i systemets PDF-läsare — där skrivardialogen bor.
         Appen har ingen egen skrivarkö och ska inte låtsas ha det. */
      window.API.json('/api/open', {
        method: 'POST', headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ path: res.path }),
      }).catch(() => {});
      window.toast && window.toast(kvittotext(res));
    });
  });
  /* Nedladdningen är motsatsen: skilda filer i en egen mapp. En lärare som
     sparar undan lektionens material vill ha tavlan, provet och facit var för
     sig — inte en enda PDF att bläddra i när hon letar efter facit. Zip valdes
     bort (ett steg till att packa upp) och likaså flera nedladdningar i rad
     (webbläsare stoppar dem som «multipla nedladdningar»). */
  $('#trycksampdf').addEventListener('click', () => {
    const b = $('#trycksampdf');
    if (!serverPa()) { window.toast && window.toast('Nedladdade — ett dokument per fil'); return; }
    bygg(b, { separat: true }, res => {
      window.API.json('/api/reveal', {
        method: 'POST', headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ path: res.path }),
      }).catch(() => {});
      window.toast && window.toast(kvittotext(res));
    });
  });

  /* Kvittot säger vad som FAKTISKT hamnade i filen — och vad som inte gjorde
     det. Ett paket som tyst blev en sida kortare upptäcks framför kopiatorn. */
  const kvittotext = res => {
    const sidor = `${res.sidor} ${res.sidor === 1 ? 'sida' : 'sidor'}`;
    const antal = (res.filer || []).length;
    const brist = (res.saknas || []).length
      ? ` ${res.saknas.join(', ')} kom inte med: ett papper som inte är godkänt har ingen PDF än.`
      : '';
    if (res.mapp) return `${antal} ${antal === 1 ? 'fil' : 'filer'} i en egen mapp — ${sidor}, mappen är öppnad.${brist}`;
    /* «I rätt ordning» sägs bara när ordningen faktiskt är hel. */
    return brist ? `Paketet är byggt — ${sidor}.${brist}`
                 : `Paketet är byggt — ${sidor} i rätt ordning.`;
  };

  function fram() {
    const r = ruta.getBoundingClientRect();
    (window.rullaTill || (y => window.scrollTo(0, y)))(Math.max(0, window.scrollY + r.top - 110));
  }

  /* Öppnas från lektionen i veckan, från veckobriefen och från planeringskön när
     lektionens papper är godkända. */
  const dagord = d => (window.Kalender && window.Kalender.ord ? window.Kalender.ord(d) : d);
  window.Tryck = {
    /* Allt som är godkänt för EN lektion — det är den hög läraren bär in. */
    lektion(post) {
      if (!post || !post.datum) return false;
      const s = (window.Dokument && window.Dokument.sparade()) || [];
      const valda = s.filter(v => v.datum === post.datum && (!post.klass || v.klass === post.klass));
      if (!valda.length) return false;
      packa([post.klass, dagord(post.datum)].filter(Boolean).join(' · '), valda);
      fram();
      return true;
    },
    oppna(namnPaPaketet) {
      const s = (window.Dokument && window.Dokument.sparade()) || [];
      const valda = namnPaPaketet
        ? s.filter(v => window.Dokument.namn(v).includes(namnPaPaketet.replace(/^Prov — /, '')))
        : s.slice(-3);
      packa(namnPaPaketet || 'nästa lektion', valda.length ? valda : s.slice(-2));
      fram();
    }
  };
})();

File: app/tryck.py
"""Utskriftspaketet (Etapp 0.9).

«Det här ska skrivas ut» som en enda gest: tavlan överst, elevernas papper
under, facit sist — i rätt antal kopior. Knappen räknade ihop högen och sa
sedan «Utskrivet» efter niohundra millisekunder. Det här bygger den på riktigt.

Paketet är EN PDF, och kopiorna ligger i den. Det är inte en omväg runt
skrivardialogen utan hela poängen: en lärare som ska ha 22 elevark, 1 tavla och
1 facit kan inte säga det i en dialog som bara har ett kopieantal för hela
jobbet. Ligger kopiorna i filen är högen redan rätt när den kommer ur skrivaren.

NEDLADDNINGEN är den andra gesten och har motsatt form: skilda filer i en egen
mapp (``dela_upp``). Läraren som sparar undan lektionens material letar efter
facit, inte efter sida 47 i en bunt.

Källorna är olika för olika papper, och det är därför den här modulen finns:

* **Prov, arbetsblad och gruppuppgifter** har redan en PDF — den som byggdes
  vid godkännandet. Den är numera en bild av SKÄRMEN: klienten ritar av varje
  blad (app/web/ui/blad-bild.js) och rutten lägger bilderna på A4 här nere
  (``png_till_pdf``, samma väg som tavlan). Tectonic är kvar som reserv när
  godkännandet kommer utan bilder, och för bedömningsanvisningen. Den tas som
  den är.
* **Facit och bedömningsanvisning** ligger bredvid provet med samma stam.
* **Tavlan** finns bara som en ritad sida i webbläsaren. Klienten skickar den
  som PNG (samma bild som /api/planning/export sparar) och den läggs på ett A4
  här — utan LaTeX, se ``png_till_pdf``.
* **Den anpassade kopian** renderas om ur provets egen JSON med längre tid,
  färre uppgifter och en dokumentkod i foten. Ingen etikett, ingen text på
  pappret som säger att det är en anpassning — koden i foten är det enda som
  skiljer den, precis som i planeringen (app/web/ui/tryck.js).
"""
from __future__ import annotations

import base64
import io
import re
import shutil
from pathlib import Path

from app import exam_gen, exam_latex, exam_pdf, exam_spec

_PNG_MAGIC = b"\x89PNG\r\n\x1a\n"
_DATA_PREFIX = "data:image/png;base64,"
MAX_KOPIOR = 60          # 22 elevark är normalt; 60 är taket mot skrivfel
MAX_PNG_BYTES = 30 * 1024 * 1024
# Taket i PIXLAR, inte bara i byte: en PNG på några kilobyte kan packa upp till
# gigabyte (en «dekompressionsbomb»). Tavlan i 2× är omkring 3 Mpx.
MAX_PIXLAR = 60_000_000
# Ett lösningsförslag är två–fyra ark. Taket är mot skrivfel och mot en klient
# som skickar en hel bok, inte mot lärarens papper.
MAX_SIDOR = 40
# A4 i punkter (72 dpi) — PDF:ens eget mått. Tavlan läggs på sidan, inte
# tvärtom: en sida i bildens storlek går inte att stoppa i en skrivare.
A4_PT = (595.276, 841.89)
# Utskriftsmarginalen. Bilden skalades förut mot PAPPERSKANTEN, och läraren såg
# det direkt när hon skrev ut på riktigt: rubriken låg dikt an vänsterkanten och
# skrivaren klippte den. En hemmaskrivare kan inte skriva i de yttersta
# millimetrarna, och ett papper utan luft ser billigt ut även när det trycks. 12
# mm är luft nog utan att göra texten liten.
MARGINAL_PT = 12 * 72 / 25.4        # ≈ 34 pt


def _safe(namn: str, fallback: str = "utskrift") -> str:
    rent = re.sub(r'[<>:"/\\|?*\x00-\x1f]', "", namn or "").strip().strip(".")
    return rent[:80] or fallback


def _platta(bild):
    """Bilden mot vitt, i RGB. Genomskinligt blir SVART i en PDF utan
    alfakanal, och tavlans mellanrum mellan två bräden är just genomskinligt
    (samma fälla som tavla-bild.js fyller duken vit för)."""
    if "A" in bild.getbands():
        from PIL import Image
        rgba = bild.convert("RGBA")
        botten = Image.new("RGB", rgba.size, "white")
        botten.paste(rgba, mask=rgba.getchannel("A"))
        return botten
    return bild if bild.mode == "RGB" else bild.convert("RGB")


def _oppna_png(dataurl: str):
    """En data:-URI till en Pillow-bild, eller None. Allt som inte är en PNG
    faller här: fel prefix, trasig base64, fel magiska byte, för stor fil."""
    if not isinstance(dataurl, str) or not dataurl.startswith(_DATA_PREFIX):
        return None
    try:
        rå = base64.b64decode(dataurl[len(_DATA_PREFIX):], validate=True)
    except (ValueError, TypeError):
        return None
    if not rå.startswith(_PNG_MAGIC) or len(rå) > MAX_PNG_BYTES:
        return None
    from PIL import Image, UnidentifiedImageError
    try:
        bild = Image.open(io.BytesIO(rå))
        # Måtten läses ur huvudet och prövas FÖRE avkodningen: några kilobyte
        # PNG kan packa upp till gigabyte, och `load()` nedan är det som
        # betalar. (Pillow har en egen bomb-vakt vid 179 Mpx som slår redan i
        # `open` — den kastar DecompressionBombError och fångas här.)
        if bild.width * bild.height > MAX_PIXLAR:
            return None
        bild.load()
    except (UnidentifiedImageError, OSError, ValueError,
            Image.DecompressionBombError):
        # Rätt magiska byte men trasig resten — klienten har ritat av något
        # annat än tavlan, och det ska sägas, inte sparas.
        return None
    return _platta(bild)


def bladbilder(blad, nyckel: str) -> list[str]:
    """Bladbilderna klienten skickade med godkännandet, ett arkläge i taget.

    Formen är ``{"uppgift": [png, …], "facit": [png, …]}`` — uppgiftsbladen och
    facit-/lösningsarken var för sig, för de blir två filer (se routes_exam
    approve). Allt som inte är en icke-tom sträng faller bort HÄR i stället för
    nere i ``png_till_pdf``: en klient som skickar tal, listor i listor eller
    null ska mötas av «inga bilder» och LaTeX-reserven, inte av ett halvbyggt
    papper. Tom lista betyder därför alltid samma sak — gå den gamla vägen.
    """
    if not isinstance(blad, dict):
        return []
    rå = blad.get(nyckel)
    if not isinstance(rå, list):
        return []
    return [b for b in rå if isinstance(b, str) and b]


def png_till_pdf(dataurl, ut_dir: Path, stam: str) -> Path | None:
    """Bilder på var sitt A4, med utskriftsmarginal. Returnerar None om någon
    inte är en PNG.

    Orienteringen väljs per bild (se nedan) och marginalen är alltid
    ``MARGINAL_PT`` — pappret ska gå att skriva ut, inte bara att titta på.

    Ingen LaTeX. Sidan byggdes förut genom att skriva PNG:en till disk och
    köra Tectonic på ett ``\\includegraphics`` — vilket band tavlans
    nedladdning till en 100 MB motor som bara skulle lägga en bild på ett
    papper, och gjorde den omöjlig på en maskin där cachen inte är seedad.
    pdfium (pypdfium2, redan här för sidräkningen och hopfogningen) lägger
    bilden på sidan direkt och komprimerar den FlateDecode — alltså
    FÖRLUSTFRITT. Det är inte en detalj: en tavla är tunn handstil på vitt, och
    JPEG — som Pillows egen PDF-export hade gett — ringlar runt varje streck.

    ``dataurl`` får vara en lista. Bokens lösningsförslag är inte ETT ark utan
    flera — svarsfacit, bedömd elevlösning, nivå 3 — och de hör till EN rad i
    utskriftsrutan. Ett anrop per ark hade gjort raden till tre rader i
    kvittot och tre filer i separatmappen för något läraren ser som ett
    papper. Alla eller ingen: en enda trasig bild fäller hela raden, för ett
    lösningsförslag som tyst tappade sin andra sida upptäcks i klassrummet.
    """
    lista = [dataurl] if isinstance(dataurl, str) else list(dataurl or [])
    if not lista or len(lista) > MAX_SIDOR:
        return None
    bilder = [_oppna_png(d) for d in lista]
    if any(b is None for b in bilder):
        return None
    import pypdfium2 as pdfium
    ut_dir.mkdir(parents=True, exist_ok=True)
    mal = ut_dir / f"{stam}.pdf"
    doc = pdfium.PdfDocument.new()
    try:
        for bild in bilder:
            b, h = bild.size
            # ORIENTERINGEN är bildens, inte pappersfackets. Ett bräde är
            # bredare än högt, och på stående A4 blev tavlan en remsa i mitten
            # med halva pappret vitt över och under — läraren såg det och sa
            # «man skulle kunna vända den 90 grader». Sidstorleken sätts per
            # sida i PDF:en, så blandade orienteringar i samma fil är fria; en
            # skrivare vänder själv. Kvadratiskt räknas som stående: den
            # naturliga vilan för ett papper.
            papper = (A4_PT[1], A4_PT[0]) if b > h else A4_PT
            # Marginalboxen, inte papperskanten. Skalan mäts mot den — det är
            # hela skillnaden mot förut.
            ruta = (papper[0] - 2 * MARGINAL_PT, papper[1] - 2 * MARGINAL_PT)
            skala = min(ruta[0] / b, ruta[1] / h)
            sida = doc.new_page(*papper)
            objekt = pdfium.PdfImage.new(doc)
            objekt.set_bitmap(pdfium.PdfBitmap.from_pil(bild))
            # Matrisen ÄR placeringen: pdfium ritar bilden i enhetskvadraten,
            # så matrisen bär både storleken och hörnet. Största möjliga med
            # bevarad proportion, centrerad i marginalboxen — som av sig själv
            # är centrerad på pappret.
            objekt.set_matrix(pdfium.PdfMatrix(
                b * skala, 0, 0, h * skala,
                (papper[0] - b * skala) / 2, (papper[1] - h * skala) / 2))
            sida.insert_obj(objekt)
            sida.gen_content()
        doc.save(str(mal))
    finally:
        doc.close()
    return mal


def anpassad_pdf(exam: dict, typ: str, ut_dir: Path, stam: str, *,
                 tid_min: int | None = None, antal: int | None = None,
                 kod: str = "") -> Path | None:
    """Den anpassade kopian: samma prov, längre tid, färre uppgifter.

    Uppgifterna som tas bort är de SISTA — provet är skrivet med stigande
    svårighet (exam_spec balanserar så), och den som får färre uppgifter ska
    få de första, inte ett slumpurval ur helheten.
    """
    if not exam_pdf.engine_available():
        return None
    kopia = exam_gen._repair_ctrl_chars({**exam})
    if antal and antal > 0:
        kopia["uppgifter"] = (kopia.get("uppgifter") or [])[:antal]
    if tid_min:
        kopia["tid_min"] = int(tid_min)
    doc, fel = exam_spec.validate_exam_json(kopia, typ)
    if doc is None:
        # Ett trimmat prov kan falla på balanskraven (färre uppgifter ändrar
        # fördelningen). Kopian är ändå lärarens beslut — rendera den utan
        # valideringen hellre än att tyst utelämna den ur paketet.
        doc, _ = exam_spec.validate_exam_json(kopia, "arbetsblad")
    if doc is None:
        return None
    tex = (exam_latex.render_arbetsblad(doc, dokumentkod=kod)
           if typ in ("arbetsblad", "gruppuppgift")
           else exam_latex.render_prov(doc, dokumentkod=kod))
    ut_dir.mkdir(parents=True, exist_ok=True)
    pdf, _logg = exam_pdf.compile_pdf(tex, ut_dir, stam)
    return pdf


def _sidor(pdf: Path) -> int:
    import pypdfium2 as pdfium
    doc = pdfium.PdfDocument(str(pdf))
    try:
        return len(doc)
    finally:
        doc.close()


def foga_ihop(delar: list[tuple[Path, int]], ut: Path) -> int:
    """Slår ihop PDF:erna i ordning, varje del upprepad sina kopior gånger.
    Returnerar sidantalet. Kopiorna ligger i FILEN — det är därför högen är
    rätt när den kommer ur skrivaren."""
    import pypdfium2 as pdfium
    paket = pdfium.PdfDocument.new()
    oppna = []
    try:
        for pdf, kopior in delar:
            src = pdfium.PdfDocument(str(pdf))
            oppna.append(src)
            sidor = list(range(len(src)))
            for _ in range(max(1, min(MAX_KOPIOR, int(kopior or 1)))):
                paket.import_pages(src, sidor)
        ut.parent.mkdir(parents=True, exist_ok=True)
        paket.save(str(ut))
        antal = len(paket)
    finally:
        for d in oppna:
            d.close()
        paket.close()
    return antal


def dela_upp(delar: list[tuple[Path, str]], mapp: Path) -> list[str]:
    """Varje dokument som EGEN fil i en egen mapp — nedladdningens form.

    Utskriften är en hopfogad hög, för det är så papperen ska komma ur
    skrivaren. Nedladdningen är motsatsen: läraren som sparar undan lektionens
    material vill ha tavlan, provet och facit som skilda filer att lägga i sin
    egen mapp — inte en enda PDF att bläddra i när hon letar efter facit.

    Kopieantalet följer INTE med. Tjugotvå exemplar av samma fil i en mapp är
    tjugoen filer för mycket; kopiorna hör hemma i högen (`foga_ihop`), där de
    faktiskt kommer ut ur maskinen.

    Numret först i filnamnet är högens ordning. En mapp sorteras alfabetiskt,
    och utan numret hamnar facit före provet — precis den ordning paketet finns
    till för att undvika. Returnerar filnamnen, i ordning, till kvittot.
    """
    mapp.mkdir(parents=True, exist_ok=True)
    filnamn: list[str] = []
    for i, (pdf, namn) in enumerate(delar, 1):
        fil = mapp / f"{i:02d} {_safe(namn, f'dokument {i}')}.pdf"
        shutil.copyfile(pdf, fil)
        filnamn.append(fil.name)
    return filnamn


def _bredvid(pdf: Path, andelse: str) -> Path | None:
    """Systerdokumentet med samma stam. Regeln bor HÄR därför att tre ställen
    behöver den — paketet, nedladdningsrutterna (routes_exam) och raderingen —
    och tre kopior av en namnregel är tre ställen att glömma."""
    kandidat = pdf.with_name(f"{pdf.stem} - {andelse}{pdf.suffix}")
    return kandidat if kandidat.is_file() else None


def bedomning_bredvid(pdf: Path) -> Path | None:
    """Bedömningsanvisningen ligger bredvid provet med samma stam
    (routes_exam approve). Finns den inte är den inte byggd — och då ska den
    inte tyst utelämnas utan saknas synligt i kvittot."""
    return _bredvid(pdf, "bedomning")


def facit_bredvid(pdf: Path) -> Path | None:
    """Arbetsbladets separata facit, samma regel som bedömningen. Bladet bär
    facit på sista sidan ändå — den här filen är för läraren som valde att ha
    lösningarna på ett eget papper."""
    return _bredvid(pdf, "facit")


def losningar_bredvid(pdf: Path) -> Path | None:
    """Provets lösningsförslag — det ark läraren SER när växlaren står på
    facitläget, avritat vid godkännandet (blad-bild.js `dokument`).

    Fram till nu pekade «Lösningar» i Sparat och i tryckpaketet på
    bedömningsanvisningen: lärarens LaTeX-satta rättningsdokument med
    kravgränser och kommenterade elevlösningar. Det är ett annat papper än det
    på skärmen, och läraren bad om skärmens.

    Anvisningen är INTE borta — den byggs som förut bredvid provet
    (``{stam} - bedomning.pdf``, routes_exam approve) och har sin egen rutt.
    Den är också reserven här: kommer godkännandet utan bilder (API-anrop,
    pytest, en äldre klient) finns ingen ``- losningar.pdf`` att hämta, och
    då ska knappen ge det dokument som faktiskt bär lösningarna i stället för
    ett 404 på ett prov som byggts felfritt."""
    return _bredvid(pdf, "losningar") or _bredvid(pdf, "bedomning")
